File: FastMines_core/src/fmg/core/mosaic/CellFill.java
package fmg.core.mosaic;

import java.beans.PropertyChangeListener;
import java.util.HashMap;
import java.util.Map;

import fmg.common.Color;
import fmg.common.notifier.INotifyPropertyChanged;
import fmg.common.notifier.NotifyPropertyChanged;
import fmg.core.types.Property;

/** all that apply to the background fill of cells */
public class CellFill implements AutoCloseable, INotifyPropertyChanged {

    public static final String PROPERTY_MODE = "Mode";

    /** режим заливки фона ячеек */
    @Property(PROPERTY_MODE)
    private int mode = 0;

    /** кэшированные цвета фона ячеек
     * <br/> Нет цвета? - создасться с нужной интенсивностью! */
    private final Map<Integer, Color> colors = new HashMap<>();

    protected NotifyPropertyChanged notifier = new NotifyPropertyChanged(this);

    /** режим заливки фона ячеек */
    public int getMode() { return mode; }

    /** режим заливки фона ячеек
     * @param newFillMode
     *  <li> 0 - цвет заливки фона по-умолчанию
     *  <li> not 0 - радуга %)
     */
    public void setMode(int newFillMode) {
        if (notifier.setProperty(mode, newFillMode, PROPERTY_MODE))
            colors.clear();
    }

    /** кэшированные цвета фона ячеек
     * <br/> Нет цвета? - создасться с нужной интенсивностью! */
    public Color getColor(int index) {
        Color color = colors.get(index);
        if (color == null) {
            color = Color.RandomColor().brighter(0.45);
            colors.put(index, color);
        }
        return color;
    }

    @Override
    public void close() {
        notifier.close();
        colors.clear();
    }

    @Override
    public void addListener(PropertyChangeListener listener) {
        notifier.addListener(listener);
    }
    @Override
    public void removeListener(PropertyChangeListener listener) {
        notifier.removeListener(listener);
    }

}

File: FastMines_core/src/fmg/core/mosaic/CellFill.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using Fmg.Common;
using Fmg.Common.Notifier;

namespace Fmg.Core.Mosaic {

    /// <summary> all that apply to the background fill of cells </summary>
    public class CellFill : INotifyPropertyChanged, IDisposable {
        /// <summary> режим заливки фона ячеек </summary>
        private int _mode = 0;
        /// <summary> кэшированные цвета фона ячеек </summary>
        private readonly IDictionary<int, Color> _colors = new Dictionary<int, Color>();

        public event PropertyChangedEventHandler PropertyChanged {
            add    { _notifier.PropertyChanged += value;  }
            remove { _notifier.PropertyChanged -= value;  }
        }
        protected readonly NotifyPropertyChanged _notifier;

        public CellFill() {
            _notifier = new NotifyPropertyChanged(this);
        }

        /// <summary> режим заливки фона ячеек
        ///  @param mode
        ///   <li> 0 - цвет заливки фона по-умолчанию
        ///   <li> not 0 - радуга %)
        /// </summary>
        public int Mode {
            get { return _mode; }
            set {
                if (_notifier.SetProperty(ref _mode, value))
                    _colors.Clear();
            }
        }

        /// <summary> кэшированные цвета фона ячеек
        /// Нет цвета? - создасться с нужной интенсивностью! */
        /// </summary>
        public Color GetColor(int index) {
            if (_colors.ContainsKey(index))
                return _colors[index];

            var res = Color.RandomColor().Brighter(0.45);
            _colors.Add(index, res);
            return res;
        }

        public void Dispose() {
            _notifier.Dispose();
            _colors.Clear();
        }

    }

}
